### PR TITLE
[compiler-dev] Compile time type deduction

### PR DESCRIPTION
**Pull request by imkiva**

### CODE REVIEW

**Comment by ice1000** [comment]
Can you provide some review instructions?

File: compiler/include/covscript/compiler/parser.hpp [hunk 1]
```diff
@@ -4,9 +4,11 @@
 #pragma once
 
 #include <antlr4-runtime.h>
-#include <covscript/compiler/sharedTypes.h>
+#include <covscript/compiler/sharedTypes.hpp>
 #include <CovScriptParser.h>
 #include <CovScriptLexer.h>
+#include <CovScriptBaseListener.h>
+#include <CovScriptBaseVisitor.h>
```
Comment (mikecovlee) [inline compiler/include/covscript/compiler/parser.hpp:11]: Consider renaming all `.h` files into `.hpp` files.

File: compiler/src/covscript/compiler/typeDeduce.cpp [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+
+        Type TypeDeduce::deduceAdditiveExpr(Parser::AdditiveExpressionContext *expr) {
+            if (expr == nullptr) {
+                return unknownType();
+            }
+
+            Type heading = deduceMultiplicativeExpr(expr->multiplicativeExpression());
+
+            if (heading.isPrimitiveType()) {
+                auto &&restExpr = expr->additiveExpressionRest();
```
Comment (mikecovlee) [inline compiler/src/covscript/compiler/typeDeduce.cpp:107]: please use move semantics instead.

File: compiler/src/covscript/compiler/typeDeduce.cpp [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+
+        Type TypeDeduce::deduceMultiplicativeExpr(Parser::MultiplicativeExpressionContext *expr) {
+            if (expr == nullptr) {
+                return unknownType();
+            }
+
+            Type heading = deduceUnaryExpr(expr->unaryExpression());
+
+            if (heading.isPrimitiveType()) {
+                auto &&restExpr = expr->multiplicativeExpressionRest();
```
Comment (mikecovlee) [inline compiler/src/covscript/compiler/typeDeduce.cpp:147]: ditto

File: compiler/src/covscript/compiler/typeDeduce.cpp [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+            if (castExpr != nullptr) {
+                return deduceCastExpr(castExpr);
+            }
+
+            auto primaryExpr = expr->primaryExpression();
+            if (primaryExpr != nullptr) {
+                return deducePrimaryExpr(primaryExpr);
+            }
+
+            return Type(TypeFlags::TYPE_UNKNOWN);
```
Comment (mikecovlee) [inline compiler/src/covscript/compiler/typeDeduce.cpp:205]: Consider using `constexpr` instead. I will tell you how.

File: compiler/src/covscript/compiler/typeDeduce.cpp [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+        }
+
+        Type TypeDeduce::deduceCastExpr(Parser::CastExpressionContext *expr) {
+            if (expr == nullptr) {
+                return unknownType();
+            }
+
+            // TODO: discover user-defined types
+            return anyType();
```
Comment (mikecovlee) [inline compiler/src/covscript/compiler/typeDeduce.cpp:214]: Ditto. Consider replacing macro with `constexpr`.

File: compiler/src/covscript/compiler/typeDeduce.cpp [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+
+        Type TypeDeduce::deducePrimaryExpr(Parser::PrimaryExpressionContext *expr) {
+            if (expr == nullptr) {
+                return unknownType();
+            }
+
+            Type heading = deducePrimaryPrefix(expr->primaryPrefix());
+
+            if (!heading.isAnyType()) {
+                auto &&restExpr = expr->primarySuffix();
```
Comment (mikecovlee) [inline compiler/src/covscript/compiler/typeDeduce.cpp:225]: Same.

**Review by mikecovlee — COMMENTED** [review]
In `variable.hpp`, try to mark the constructor of `Type` a `constexpr`. In this way you can replace some macros.
```c++
constexpr explicit Type(TypeFlag typeFlag);
```

File: compiler/src/covscript/compiler/typeDeduce.cpp (hunk 2 above)
Comment (imkiva) [reply compiler/src/covscript/compiler/typeDeduce.cpp:107]: I will try.

File: compiler/include/covscript/compiler/parser.hpp (hunk 1 above)
Comment (imkiva) [reply compiler/include/covscript/compiler/parser.hpp:11]: I will try.

File: compiler/src/covscript/compiler/typeDeduce.cpp (hunk 3 above)
Comment (imkiva) [reply compiler/src/covscript/compiler/typeDeduce.cpp:147]: ditto

File: compiler/src/covscript/compiler/typeDeduce.cpp (hunk 4 above)
Comment (imkiva) [reply compiler/src/covscript/compiler/typeDeduce.cpp:205]: dio

File: compiler/src/covscript/compiler/typeDeduce.cpp (hunk 5 above)
Comment (imkiva) [reply compiler/src/covscript/compiler/typeDeduce.cpp:214]: asgard!

File: compiler/src/covscript/compiler/typeDeduce.cpp (hunk 6 above)
Comment (imkiva) [reply compiler/src/covscript/compiler/typeDeduce.cpp:225]: rainbow bridge.

**Comment by mikecovlee** [comment]
I am still pretty annoyed by the lack of type-check in `antlr::Any`. Why not replace it with `cs::shared::any`?
**This is just a suggestion.**

**Comment by imkiva** [comment]
> I am still pretty annoyed by the lack of type-check in `antlr::Any`. Why not replace it with `cs::shared::any`?
> **This is just a suggestion.**

If type-check is really needed, we will consider replacing it at that time.

**Pull request merged**